toolchain/check/handle_impl.cpp

```diff
@@ -502,7 +505,7 @@ static auto BuildImplDecl(Context& context, Parse::AnyImplDeclId node_id,
   ReplaceInstBeforeConstantUse(context, impl_decl_id, impl_decl);
 
   // For an `extend impl` declaration, mark the impl as extending this `impl`.
-  if (self_type_id != SemIR::ErrorInst::TypeId &&
+  if (self_type_inst_id != SemIR::ErrorInst::TypeInstId &&
```

Comment:
I think still want to look at `GetTypeIdForTypeInstId(self_type_inst_id)` here, in order to handle the case where we have a `Self` type expression but it contains an error. (The `Self` could have an error nested within it without being an error instruction.) For example:
```carbon
interface I {}
impl nonexistent* as I {}
```

Reply:
Okay, added a test for that case. I think it does prevent an extra diagnostic with the current implementation. Does the current factoring still make sense to you to proceed with?